### PR TITLE
Home8

### PR DESCRIPTION
**Pull request by IgorPrihodko**

### CODE REVIEW

**Review by IgorPrihodko — COMMENTED** [review]
Sorry for files from homework7, ill fixed it in future.

File: src/main/java/homework8/ArrayListImplementation.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package homework8;
+
+import java.util.List;
+
+public class ArrayListImplementation<T> {
+
+    private T[] typeArr;
+    private int size;
+
+    public ArrayListImplementation() {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java:12]: by default, this constructor constructs an empty list with an initial capacity of ten.

File: src/main/java/homework8/ArrayListImplementation.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    int size() {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: maybe better write like this:
return size;

add access modifier

File: src/main/java/homework8/ArrayListImplementation.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        return size;
+    }
+
+    boolean isEmpty() {
+        return size == 0;
+    }
+
+    void add(T value) {
+        T[] newArr = (T[]) new Object[size + 1];
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: we need increase size of the array in 1.5
(10->15->...)

File: src/main/java/homework8/ArrayListImplementation.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        System.arraycopy(typeArr, 0, newArr, 0, typeArr.length);
+        typeArr = newArr;
+        typeArr[size++] = value;
+    }
+
+    void add(T value, int index) {
+        T[] newArr = (T[]) new Object[size + 1];
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: list = new ArrayListImplementation();
list.add(2, 100); or list.add(2, -5); 
test what happens here?

File: src/main/java/homework8/ArrayListImplementation.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package homework8;
+
+import java.util.List;
+
+public class ArrayListImplementation<T> {
+
+    private T[] typeArr;
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: ArrayListImplementation <Integer> l = new ArrayListImplementation<>();
l.size();
test what happens

File: src/main/java/homework8/ArrayListImplementation.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    void addAll(List<T> list) {
+        T[] newArr = (T[]) new Object[size + list.size()];
+        System.arraycopy(typeArr, 0, newArr, 0, typeArr.length);
+        System.arraycopy(list, 0, newArr, size + 1, list.size());
+        typeArr = newArr;
+    }
+
+    T get(int index) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: better add checking: if (index < 0&& index>size);

File: src/main/java/homework8/ArrayListImplementation.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        System.arraycopy(typeArr, 0, newArr, 0, index);
+        System.arraycopy(typeArr, index + 1, newArr, index, size - index - 1);
+        typeArr = newArr;
+        return value;
+    }
+
+    T remove(T t) {
+        for (int i = 0; i < size; i++) {
+            T value = typeArr[i];
+            if (typeArr[i] == t) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: try to use method remove(int index) to simplify code

File: src/main/java/homework8/ArrayListImplementation.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+public class ArrayListImplementation<T> {
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+    private T[] typeArr;
+    private int size;
+
+    public ArrayListImplementation(int initialCapacity) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java:16]: add other constructor without parameter which will constructs an empty list with an initial capacity of ten

File: src/main/java/homework8/ArrayListImplementation.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
+package homework8;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class ArrayListImplementation<T> {
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+    private T[] typeArr;
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: better rename to arrayOfValues of create other name

File: src/main/java/homework8/ArrayListImplementation.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            throw new IllegalArgumentException("Wrong capacity: " + initialCapacity);
+        }
+    }
+
+    private int size() {
+        return size;
+    }
+
+    private boolean isEmpty() {
+        return this.typeArr == EMPTY_ELEMENTDATA;
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: better write: return size==0

File: src/main/java/homework8/ArrayListImplementation.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    private void add(T value) {
+        if (isEmpty() || size == typeArr.length) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: try to simplify:
if (isEmpty() || size == typeArr.length) {
            T[] newArr = (T[]) new Object[(typeArr.length * 3) / 2 + 1];
            System.arraycopy(typeArr, 0, newArr, 0, typeArr.length);
            typeArr = newArr;
        } 
            typeArr[size] = value;
        size++;

File: src/main/java/homework8/ArrayListImplementation.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    private void add(T value, int index) {
+        if (index < 0) {
+            System.out.println("Wrong index. Need index >= 0");
+            return;
+        }
+        if (isEmpty()) {
+            typeArr = (T[]) new Object[index + 1];
+            typeArr[index] = value;
+        } else if (index <= typeArr.length) {
+            T[] newArr = (T[]) new Object[(typeArr.length * 3) / 2 + 1];
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: for this purpose better add checking: if(size>=typeArr.length)

File: src/main/java/homework8/ArrayListImplementation.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+            System.arraycopy(typeArr, 0, newArr, 0, index);
+            System.arraycopy(typeArr, index, newArr, index + 1, typeArr.length - index);
+            typeArr = newArr;
+            typeArr[index] = value;
+        } else if (index > typeArr.length) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: do we need this checking here?

File: src/main/java/homework8/ArrayListImplementation.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            T[] newArr = (T[]) new Object[index + 1];
+            System.arraycopy(typeArr, 0, newArr, 0, typeArr.length);
+            typeArr = newArr;
+            typeArr[index] = value;
+        }
+        size++;
+    }
+
+    private void addAll(List<T> list) {
+        T[] newArr = (T[]) new Object[typeArr.length + list.size()];
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: the same about checking

File: src/main/java/homework8/ArrayListImplementation.java (hunk 6 above)
Comment (Sofasmile) [reply src/main/java/homework8/ArrayListImplementation.java]: resolve this comments)

File: src/main/java/homework8/ArrayListImplementation.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        typeArr = newArr;
+        size = size + list.size();
+    }
+
+    T get(int index) {
+        return typeArr[index];
+    }
+
+    public void set(T value, int index) {
+        if (index >= 0) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: add if(index > typeArr.length)

File: src/main/java/homework8/ArrayListImplementation.java [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    private int size() {
+        return size;
+    }
+
+    private boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void add(T value) {
+        if (isEmpty()) {
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java:37]: we don't need this if statement here

File: src/main/java/homework8/ArrayListImplementation.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+            size = size + list.size();
+        } else {
+            for (T value: list) {
+                add(value);
+            }
+        }
+    }
+
+    public T get(int index) {
+        return arrayOfValues[index];
```
Comment (Sofasmile) [inline src/main/java/homework8/ArrayListImplementation.java]: list = new ArrayListImplementation();
list.get(-5);
test what happens here?

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**